src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/Utils/TypeNameHelpers.cs

```diff
@@ -0,0 +1,187 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Reflection.Metadata;
+using System.Runtime.CompilerServices;
+using System.Runtime.Serialization;
+using System.Text;
+
+namespace System.Formats.Nrbf.Utils;
+
+internal static class TypeNameHelpers
+{
+    // PrimitiveType does not define Object, IntPtr or UIntPtr
+    internal const PrimitiveType ObjectPrimitiveType = (PrimitiveType)19;
+    internal const PrimitiveType IntPtrPrimitiveType = (PrimitiveType)20;
+    internal const PrimitiveType UIntPtrPrimitiveType = (PrimitiveType)21;
+    private static readonly TypeName?[] s_PrimitiveTypeNames = new TypeName?[(int)UIntPtrPrimitiveType + 1];
+    private static readonly TypeName?[] s_PrimitiveSZArrayTypeNames = new TypeName?[(int)UIntPtrPrimitiveType + 1];
+    private static AssemblyNameInfo? s_CoreLibAssemblyName;
+
+    internal static TypeName GetPrimitiveTypeName(PrimitiveType primitiveType)
+    {
+        Debug.Assert(primitiveType is not (PrimitiveType.None or PrimitiveType.Null));
+
+        TypeName? typeName = s_PrimitiveTypeNames[(int)primitiveType];
+        if (typeName is null)
+        {
+            string fullName = primitiveType switch
+            {
+                PrimitiveType.Boolean => "System.Boolean",
+                PrimitiveType.Byte => "System.Byte",
+                PrimitiveType.SByte => "System.SByte",
+                PrimitiveType.Char => "System.Char",
+                PrimitiveType.Int16 => "System.Int16",
+                PrimitiveType.UInt16 => "System.UInt16",
+                PrimitiveType.Int32 => "System.Int32",
+                PrimitiveType.UInt32 => "System.UInt32",
+                PrimitiveType.Int64 => "System.Int64",
+                PrimitiveType.UInt64 => "System.UInt64",
+                PrimitiveType.Single => "System.Single",
+                PrimitiveType.Double => "System.Double",
+                PrimitiveType.Decimal => "System.Decimal",
+                PrimitiveType.TimeSpan => "System.TimeSpan",
+                PrimitiveType.DateTime => "System.DateTime",
+                PrimitiveType.String => "System.String",
+                ObjectPrimitiveType => "System.Object",
+                IntPtrPrimitiveType => "System.IntPtr",
+                UIntPtrPrimitiveType => "System.UIntPtr",
+                _ => throw new ArgumentOutOfRangeException(paramName: nameof(primitiveType), actualValue: primitiveType, message: null)
```

Comment:
I'm on the fence whether I like this lazy initialization because it shows the symbolic name to value matching, or if it should just be declared as `private static readonly string?[] s_primitiveTypeNames = { "System.Boolean", "System.Byte", ..., null, null, "System.SomethingElse", ... };`

Reply:
Oh, right, this then takes the string and does further work (TypeName.Parse) before squirrelling it away in the array.

OK, this is probably the better approach.